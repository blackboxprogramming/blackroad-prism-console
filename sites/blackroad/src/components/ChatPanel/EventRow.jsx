import React from 'react';

export default function EventRow({ event }) {
  return (
    <div className="rounded border border-dashed border-slate-300 bg-slate-50 px-3 py-2 text-xs text-slate-600">
      <div className="flex items-center justify-between">
        <span className="font-semibold uppercase tracking-wide">{event.role}</span>
        <span>{new Date(event.ts).toLocaleTimeString()}</span>
      </div>
      <p className="mt-1 whitespace-pre-wrap text-slate-600">{event.text}</p>
    </div>
  );
}
