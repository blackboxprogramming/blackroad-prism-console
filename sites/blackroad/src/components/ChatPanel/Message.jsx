import React from 'react';

export default function Message({ message }) {
  const attachments = message.attachments ?? [];
  const reactions = message.reactions ?? {};
  const redactions = message.redactions ?? [];

  return (
    <article className="rounded border border-slate-200 bg-white p-3 shadow-sm">
      <header className="mb-2 flex items-center justify-between text-xs">
        <span className="font-semibold text-slate-800">{message.author}</span>
        <span className="text-slate-400">{new Date(message.ts).toLocaleString()}</span>
      </header>
      <p className="whitespace-pre-wrap text-sm text-slate-800">{message.text}</p>
      {attachments.length ? (
        <ul className="mt-3 space-y-2">
          {attachments.map((attachment, index) => (
            <li key={index} className="text-xs">
              <a
                href={attachment.url}
                target="_blank"
                rel="noreferrer"
                className="rounded border border-slate-200 bg-slate-50 px-2 py-1 font-medium text-slate-700 hover:bg-slate-100"
              >
                {attachment.kind.toUpperCase()} Â· {attachment.url}
              </a>
            </li>
          ))}
        </ul>
      ) : null}
      {Object.keys(reactions).length ? (
        <div className="mt-3 flex flex-wrap gap-2 text-xs text-slate-500">
          {Object.entries(reactions).map(([emoji, count]) => (
            <span key={emoji} className="rounded bg-slate-100 px-2 py-1">
              {emoji} {count}
            </span>
          ))}
        </div>
      ) : null}
      {redactions.length ? (
        <p className="mt-3 rounded bg-amber-50 px-2 py-1 text-xs text-amber-700">
          Redactions applied: {redactions.join(', ')}
        </p>
      ) : null}
    </article>
  );
}
