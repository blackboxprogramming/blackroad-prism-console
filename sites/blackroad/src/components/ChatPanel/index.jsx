import React, { useEffect, useMemo, useRef, useState } from 'react';
import { connect } from '@blackroad/chat-sdk';
import Message from './Message.jsx';
import EventRow from './EventRow.jsx';

const DEFAULT_MESSAGES = [];

export default function ChatPanel({ jobId, protocol = 'graphql', baseUrl = '', permissions, initialMessages }) {
  const [messages, setMessages] = useState(initialMessages ?? DEFAULT_MESSAGES);
  const [input, setInput] = useState('');
  const [status, setStatus] = useState('idle');
  const [error, setError] = useState(null);
  const clientRef = useRef(null);
  const listRef = useRef(null);

  const canPost = permissions?.canPost ?? true;

  useEffect(() => {
    if (!jobId) {
      setMessages(initialMessages ?? DEFAULT_MESSAGES);
      return () => {};
    }

    const client = connect({ jobId, protocol, baseUrl });
    clientRef.current = client;
    let mounted = true;
    setStatus('connecting');

    client
      .fetchThread()
      .then((thread) => {
        if (mounted) {
          setMessages(thread);
          setStatus('ready');
        }
      })
      .catch((err) => {
        if (mounted) {
          setStatus('error');
          setError(err);
        }
      });

    const unsubscribe = client.subscribe((message) => {
      setMessages((prev) => {
        const seen = new Map(prev.map((item) => [item.id, item]));
        seen.set(message.id, message);
        return Array.from(seen.values()).sort((a, b) => a.ts.localeCompare(b.ts));
      });
    });

    return () => {
      mounted = false;
      unsubscribe();
      client.close();
      if (clientRef.current === client) {
        clientRef.current = null;
      }
    };
  }, [jobId, protocol, baseUrl, initialMessages]);

  useEffect(() => {
    if (listRef.current) {
      listRef.current.scrollTop = listRef.current.scrollHeight;
    }
  }, [messages]);

  const handleSubmit = async (event) => {
    event.preventDefault();
    if (!input.trim() || !clientRef.current) return;
    try {
      setStatus('posting');
      await clientRef.current.post(input.trim());
      setInput('');
      setStatus('ready');
    } catch (err) {
      setStatus('error');
      setError(err);
    }
  };

  const ordered = useMemo(() => [...messages].sort((a, b) => a.ts.localeCompare(b.ts)), [messages]);

  return (
    <div className="flex h-full flex-col rounded-lg border border-slate-200 bg-white shadow-sm">
      <div className="flex items-center justify-between border-b border-slate-200 px-4 py-2">
        <div>
          <h2 className="text-sm font-semibold text-slate-900">Chat</h2>
          {jobId ? <p className="text-xs text-slate-500">Live events for job {jobId}</p> : <p className="text-xs text-slate-500">Select a job</p>}
        </div>
        <span className="text-xs uppercase text-slate-500">{status}</span>
      </div>

      {error ? <div className="px-4 py-2 text-sm text-red-600">{error.message}</div> : null}

      <div ref={listRef} className="flex-1 overflow-auto px-4 py-3">
        {!ordered.length ? <p className="text-sm text-slate-500">No events yet.</p> : null}
        <ol className="space-y-3">
          {ordered.map((message) => (
            <li key={message.id}>
              {message.role === 'system' ? <EventRow event={message} /> : <Message message={message} />}
            </li>
          ))}
        </ol>
      </div>

      <form onSubmit={handleSubmit} className="border-t border-slate-200 p-3">
        <label className="mb-2 block text-xs font-semibold text-slate-600" htmlFor="chat-input">
          Reply
        </label>
        <textarea
          id="chat-input"
          value={input}
          onChange={(event) => setInput(event.target.value)}
          disabled={!canPost || !jobId}
          placeholder={canPost ? 'Send a note, /command, or link an artifactâ€¦' : 'Read-only access'}
          className="h-24 w-full resize-none rounded border border-slate-300 bg-white p-2 text-sm text-slate-800 focus:border-slate-500 focus:outline-none"
        />
        <div className="mt-2 flex items-center justify-between text-xs text-slate-500">
          <span>{input.startsWith('/') ? 'Commands: /rerun, /cancel, /pin, /link <artifactId>' : 'Supports markdown + /commands'}</span>
          <button
            type="submit"
            disabled={!canPost || !input.trim()}
            className="rounded bg-slate-900 px-3 py-1 font-semibold text-white disabled:cursor-not-allowed disabled:opacity-60"
          >
            Send
          </button>
        </div>
      </form>
    </div>
  );
}
