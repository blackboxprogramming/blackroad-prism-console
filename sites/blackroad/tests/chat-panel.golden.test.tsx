import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';

const SAMPLE_MESSAGES = [
  {
    id: 'chat-1',
    jobId: 'demo',
    author: 'Operator',
    role: 'user',
    ts: '2025-01-15T12:00:00.000Z',
    text: 'Init job run /rerun',
    reactions: { '👍': 1 },
    attachments: [],
    redactions: [],
  },
  {
    id: 'chat-2',
    jobId: 'demo',
    author: 'Agent',
    role: 'agent',
    ts: '2025-01-15T12:05:00.000Z',
    text: 'Run completed, artifact ready.',
    reactions: {},
    attachments: [{ kind: 'image', url: '/static/frame1.png' }],
    redactions: [],
  },
];

vi.mock('@blackroad/chat-sdk', () => ({
  connect: () => ({
    protocol: 'sse',
    subscribe: () => () => {},
    fetchThread: async () => SAMPLE_MESSAGES,
    post: async () => SAMPLE_MESSAGES[0],
    close: () => {},
  }),
}));

import ChatPanel from '../src/components/ChatPanel/index.jsx';

describe('ChatPanel', () => {
  it('renders message feed with composer', async () => {
    const { container, findByText } = render(
      <ChatPanel jobId="demo" protocol="sse" initialMessages={SAMPLE_MESSAGES} permissions={{ canPost: true }} />,
    );

    await findByText('Run completed, artifact ready.');

    expect(container.firstChild).toMatchSnapshot();
  });
});
