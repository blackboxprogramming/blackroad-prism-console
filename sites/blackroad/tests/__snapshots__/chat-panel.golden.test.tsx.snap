// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatPanel > renders message feed with composer 1`] = `
<div
  class="flex h-full flex-col rounded-lg border border-slate-200 bg-white shadow-sm"
>
  <div
    class="flex items-center justify-between border-b border-slate-200 px-4 py-2"
  >
    <div>
      <h2
        class="text-sm font-semibold text-slate-900"
      >
        Chat
      </h2>
      <p
        class="text-xs text-slate-500"
      >
        Live events for job 
        demo
      </p>
    </div>
    <span
      class="text-xs uppercase text-slate-500"
    >
      ready
    </span>
  </div>
  <div
    class="flex-1 overflow-auto px-4 py-3"
  >
    <ol
      class="space-y-3"
    >
      <li>
        <article
          class="rounded border border-slate-200 bg-white p-3 shadow-sm"
        >
          <header
            class="mb-2 flex items-center justify-between text-xs"
          >
            <span
              class="font-semibold text-slate-800"
            >
              Operator
            </span>
            <span
              class="text-slate-400"
            >
              1/15/2025, 12:00:00 PM
            </span>
          </header>
          <p
            class="whitespace-pre-wrap text-sm text-slate-800"
          >
            Init job run /rerun
          </p>
          <div
            class="mt-3 flex flex-wrap gap-2 text-xs text-slate-500"
          >
            <span
              class="rounded bg-slate-100 px-2 py-1"
            >
              üëç
               
              1
            </span>
          </div>
        </article>
      </li>
      <li>
        <article
          class="rounded border border-slate-200 bg-white p-3 shadow-sm"
        >
          <header
            class="mb-2 flex items-center justify-between text-xs"
          >
            <span
              class="font-semibold text-slate-800"
            >
              Agent
            </span>
            <span
              class="text-slate-400"
            >
              1/15/2025, 12:05:00 PM
            </span>
          </header>
          <p
            class="whitespace-pre-wrap text-sm text-slate-800"
          >
            Run completed, artifact ready.
          </p>
          <ul
            class="mt-3 space-y-2"
          >
            <li
              class="text-xs"
            >
              <a
                class="rounded border border-slate-200 bg-slate-50 px-2 py-1 font-medium text-slate-700 hover:bg-slate-100"
                href="/static/frame1.png"
                rel="noreferrer"
                target="_blank"
              >
                IMAGE
                 ¬∑ 
                /static/frame1.png
              </a>
            </li>
          </ul>
        </article>
      </li>
    </ol>
  </div>
  <form
    class="border-t border-slate-200 p-3"
  >
    <label
      class="mb-2 block text-xs font-semibold text-slate-600"
      for="chat-input"
    >
      Reply
    </label>
    <textarea
      class="h-24 w-full resize-none rounded border border-slate-300 bg-white p-2 text-sm text-slate-800 focus:border-slate-500 focus:outline-none"
      id="chat-input"
      placeholder="Send a note, /command, or link an artifact‚Ä¶"
    />
    <div
      class="mt-2 flex items-center justify-between text-xs text-slate-500"
    >
      <span>
        Supports markdown + /commands
      </span>
      <button
        class="rounded bg-slate-900 px-3 py-1 font-semibold text-white disabled:cursor-not-allowed disabled:opacity-60"
        disabled=""
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;
